citar.el

```diff
@@ -885,13 +897,16 @@ into the corresponding reference key.  Return
 
 ;;;###autoload
 (defun citar-open (keys-entries)
-  "Open related resources (links or files) for KEYS-ENTRIES."
+  "Open any related resource (links, files, or notes) for KEYS-ENTRIES."
   (interactive (list (citar-select-refs
                       :rebuild-cache current-prefix-arg)))
   (when (and citar-library-paths
              (stringp citar-library-paths))
     (message "Make sure 'citar-library-paths' is a list of paths"))
-  (let* ((key-entry-alist (citar--ensure-entries keys-entries))
+  (let* ((embark-default-action-overrides '((consult-multi . citar-open-multi)
+					    (file . citar-file-open)
+                                            (url . browse-url)))
```

Comment:
Here the consult-multi and url overrides should probably go away.

Reply:
Thanks for looking at this, and for the comments.

Embark and consult-multi work just as you describe when it comes to categorizing files and urls, but (as I describe [here](https://github.com/bdarcus/citar/pull/491#issuecomment-998660728)) because the source of the target is minibuffer completion, Embark sets the default action for all targets, regardless of type, to the command that opened the minibuffer in the first place (per the docstring of `embark-default-action-overrides`). That means the default action for any target type is set to `citar open`, which is a function that does not take files or urls as arguments. Hence the overrides. It's a bit of a hail mary, I guess, but I'm afraid I'm running up against the limits of my lisp/emacs knowledge here. ;)

As for the `citar-open-multi` function, it was only added to allow `embark-act-all` to work on properly on targets of multiple types. When using `embark-act` on single candidates, consult-multi works just as expected and this function isn't called. When calling `embark-act-all` on targets of multiple types, they haven't been transformed by the candidate transformer yet ---  so they remain typed as `consult-multi`. Hence, another override. Again, bit of a hail mary -- but functional!

Thanks again for your comments. Really appreciate your taking the time.

Reply:
Okay, maybe I missed something here but this seems wrong. The default action is supposed to be *the command that started the completion*. Overriding it with something else is counterintuitive and goes against the usual Embark behavior.

Note that Embark has a minibuffer argument injection mechanism. I suggest you dig deeper into the Embark documentation.

Reply:
> That means that, in this case, the default action for any target type is set to citar open, which is a function that does not take files or urls as arguments. Hence the overrides. 

I am specifically referring to this. This is wrong. There is the minibuffer injection mechanism.

> When calling embark-act-all on targets of multiple types, they haven't been transformed by the candidate transformer yet --- so they remain typed as consult-multi.

I will check this, but if this is the case it is a bug or issue in Embark. 

I know that oantolin added some target type checking to embark-act-all and there is indeed a problem when acting on candidates of multiple categories at once. Anyways this should be handled properly by  Embark itself.

Reply:
I experimented a bit with removing these lines, and things seem to work fine, until I run `embark-act-all` on the resulting resources. So to recap:

1. `citar-open`
2. narrow candidates and then `embark-act-all`
3. from there I get a list of resources, if I run `embark-act-all`, II then get the error below:

```elisp
embark PCH: (user-error "No target found")
```

... which is I think why he introduced the `citar-open-multi` function.

![image](https://user-images.githubusercontent.com/1134/147260488-87902c82-b8aa-48aa-a75c-c1065c8bc725.png)

Reply:
Did you load embark-consult?

Reply:
> Did you load embark-consult? 

I did require it now, and now get:

```elisp
(execute-extended-command nil "citar-open" nil)
```

> I don't know what is wrong here exactly, but the override should not be needed due to the design of embark. consult-buffer also works without override.

I'll take another look.

Reply:
@bdarcus What's the status here? Did you figure out the issues?

Reply:
> Did you figure out the issues?

Not yet. Likely in the next few days.